src/nvim/decoration.c

```diff
@@ -792,50 +792,113 @@ DecorSignHighlight *decor_find_sign(DecorInline decor)
   }
 }
 
-// Increase the signcolumn size and update the sentinel line if necessary for
-// the invalidated range.
-void decor_validate_signcols(buf_T *buf, int max)
+/// If "count" is greater than current max, set it and make b_signcols valid.
+/// Adjust max_count if validated row has or had max signs.
+void buf_signcols_validate_row(buf_T *buf, int count, int add)
 {
-  int signcols = 0;  // highest value of count
-  int currow = buf->b_signcols.invalid_top - 1;
-  // TODO(bfredl): only need to use marktree_itr_get_overlap once.
-  // then we can process both start and end events and update state for each row
-  for (; currow < buf->b_signcols.invalid_bot; currow++) {
-    MarkTreeIter itr[1];
-    if (!marktree_itr_get_overlap(buf->b_marktree, currow, 0, itr)) {
-      continue;
+  int del = add < 0 ? -add : 0;
+  if (count > buf->b_signcols.max) {
+    buf->b_signcols.max = count;
+    buf->b_signcols.resized = true;
+    buf->b_signcols.max_count = 0;
+  }
+  buf->b_signcols.max_count += (count == buf->b_signcols.max - del) * (del ? -1 : 1);
+}
+
+void buf_signcols_validate_range(buf_T *buf, int row1, int row2, int add)
+{
+  int count = 0;
+  int currow = row1;
+  MTPair pair = { 0 };
+  MarkTreeIter itr[1];
+  int *overlap = xcalloc(sizeof(int), (size_t)(row2 + 1 - row1));
+
+  marktree_itr_get_overlap(buf->b_marktree, currow, 0, itr);
+  while (marktree_itr_step_overlap(buf->b_marktree, itr, &pair)) {
```

Comment:
I know this might be a common way of finding range intersections, but the API for extmarks here is a bit esoteric, so some descriptions of what is happening will make this easier to follow. I.e. why is this only done for `row1`?

Reply:
The explanation is given in the `marktree_get_overlap()` docstring. Can add a comment referring to that I guess.